src/completion.ml

```diff
@@ -0,0 +1,292 @@
+     meta-variable [m] *)
+  let to_term_env : meta option -> term_env = fun m ->
+    let m = match m with Some m -> m | None -> assert false in
+    let xs = Array.init m.meta_arity (Printf.sprintf "x%i") in
+    let xs = Bindlib.new_mvar mkfree xs in
+    let ar = Array.map _Vari xs in
+    TE_Some (Bindlib.unbox (Bindlib.bind_mvar xs (_Meta m ar))) in
+  let terms_env = Array.map to_term_env metas in
+  let rhs = Bindlib.msubst r.rhs terms_env in
+  (lhs, rhs)
+
+(** First-order syntactic unification *)
+
+exception Unsolvable
+
+(** [unif_aux t1 t2 eqs] attempts to solve the first-order unification
+    problem [(t1 = t2) :: eqs] *)
+let rec unif_aux : term -> term -> unif_constrs -> unit = fun t1 t2 eqs ->
+  match (t1, t2) with
+  | Type, Type
+  | Kind, Kind                    -> unif eqs
+  | Symb(s1, _), Symb(s2, _)      ->
+      if s1 == s2 then unif eqs else raise Unsolvable
+  | Vari _, _
+  | _, Vari _
+  | Prod _, _
+  | _, Prod _
+  | Abst _, _
+  | _, Abst _ -> assert false
+  | Appl (t1, u1), Appl (t2, u2)  -> unif ((t1, t2) :: (u1, u2) :: eqs)
+  | Meta (m1, ts1), Meta (_, ts2) ->
+      assert (ts1 = [||] && ts2 = [||]);
+      let vs = Bindlib.new_mvar mkfree [||] in
+      let bt2 = Bindlib.bind_mvar vs (lift t2) in
+      set_meta m1 (Bindlib.unbox bt2);
+      unif eqs
+  | (Meta (m1, ts1), t2)
+  | (t2, Meta (m1, ts1))          ->
+      if occurs m1 t2 then raise Unsolvable
+      else
+        assert (ts1 = [||]);
+        let vs = Bindlib.new_mvar mkfree [||] in
+        let bt2 = Bindlib.bind_mvar vs (lift t2) in
+        set_meta m1 (Bindlib.unbox bt2);
+        unif eqs
+  | _                             -> raise Unsolvable
+
+(** [unif eqs] attempts to solve the unification problem [eqs] *)
+and unif : unif_constrs -> unit = fun eqs ->
+  match eqs with
+  | []              -> ()
+  | (t1, t2) :: eqs -> unif_aux t1 t2 eqs
+
+(** [cps rs] computes the critical pairs of the rewrite system [rs] *)
+let cps : rule list -> (term * term) list = fun rs ->
+  let acc = ref [] in
+  let cps_aux : rule -> rule -> unit = fun r1 r2 ->
+    let lhs1, rhs1 = to_term r1 in
+    let lhs2, rhs2 = to_term r2 in
+    let find_cp : term -> term -> unit = fun t1 t2 ->
+      let reset_meta m = m.meta_value := None in
+      iter_meta reset_meta lhs1;
+      iter_meta reset_meta lhs2;
+      try
+        unif [(t1, t2)];
+        acc := (rhs1, rhs2) :: !acc
+      with Unsolvable -> () in
+    let _, args1 = get_args lhs1 in
+    let _, args2 = get_args lhs2 in
+    find_cp lhs1 lhs2;
+    List.iter (iter (find_cp lhs1)) args2;
+    List.iter (iter (fun t -> find_cp t lhs2)) args1 in
+  let rec cps rs =
+    match rs with
+    | []       -> ()
+    | r :: rs' -> List.iter (cps_aux r) rs; cps rs' in
+  cps rs;
+  !acc
+
+(** Rewriting for closed rules and closed terms *)
+
+type ctxt = term -> term
+
+(** [match_rule ctxt t (s, r)] attempts to rewrite [ctxt[t]] by appling the
```

Comment:
You should not write yet another normalization function but use Eval.snf instead by replacing metavariables in rules by fresh function symbols with no rule (function gamma).

Reply:
The reason why I didn't choose to use Eval.snf is that when we attempts to apply rules in Eval.find_rule, we only consider the rules stored in s.sym_rules However, in the algorithm of completion, we may need to compute the normal forms of a term in different rewrite systems. I don't think it's a good idea to update the field s.sym_ruies every time and that's why I implemented an ad hoc function for computing normal forms of closed terms.

Reply:
No, there is no problem in updating the sym_rules field. We can use the Timed module to backtrack if necessary.